packages/mobx-state-tree/__tests__/core/create-action-tracking-middleware.test.ts

```diff
@@ -0,0 +1,36 @@
+import { addMiddleware, createActionTrackingMiddleware, types, flow } from "../../src"
+
+const middleware = createActionTrackingMiddleware({
+    onStart: () => {},
+    onResume: () => {},
+    onSuspend: () => {},
+    onSuccess: () => {},
+    onFail: () => {}
+})
+
+const M = types
+    .model({})
+    .actions(self => ({
+        innerFlow: flow(function*() {
+            yield new Promise(resolve => setTimeout(resolve))
+            return true
+        })
+    }))
+    .actions(self => ({
+        outerFlow: flow(function*() {
+            yield self.innerFlow()
+            return true
+        })
+    }))
+
+test("can handle inner flow", async () => {
+    const m = M.create()
+    addMiddleware(m, middleware)
+    await expect(m.innerFlow()).resolves.toBe(true)
+})
```

Comment:
could you also check that the actionTrackingMiddlewares are properly called in the right order?
Something like having a string array and then each action tracking middleware hook pushes something to that event and doing a expect(events).toMatchInlineSnapshot

same for the other test

Reply:
Working through the suggested test changes and I think it's uncovering some more issues in the middleware from nested flow.

Just to make sure we're on the same page on expectations, given actions that look like the following:
```
innerFlow: flow(function *() {
  console.log("innerFlow - run 1")
  yield new Promise(resolve => setTimeout(resolve))
  console.log("innerFlow - run 1")
})
outerFlow: flow(function *() {
  console.log("outerFlow - run 1")
  yield self. innerFlow();
  console.log("outerFlow - run 2")
});
```

I would expect to see:
```
outerFlow - onStart
outerFlow - onResume
outerFlow - run 1
innerFlow - onStart
innerFlow - onResume
innerFlow - run 1
innerFlow - onSuspend
outerFlow - onSuspend
outerFlow - onResume
innerFlow - onResume
innerFlow - run 2
innerFlow - onSuspend
innerFlow - onSuccess
outerFlow - onSuspend
outerFlow - onResume
outerFlow - run 2
outerFlow - onSuspend
outerFlow - onSuccess
```

If the plan is to deprecate the old middleware API, would you be open to changing createActionTrackingMiddleware2 interface as follows:
```
export interface IActionTrackingMiddleware2Hooks<TEnv> {
    filter?: (call: IActionTrackingMiddleware2Call<TEnv>) => boolean
    onStart: (call: IActionTrackingMiddleware2Call<TEnv>) => void
    onSuspend?: (call: IActionTrackingMiddleware2Call<TEnv>) => void
    onResume?: (call: IActionTrackingMiddleware2Call<TEnv>) => void
    onFinish: (call: IActionTrackingMiddleware2Call<TEnv>, error?: any) => void
}
```
such that onSuspend and onResume are only necessary if a user needs those hooks (but doesn't complicate the usage for the simple cases)?

Reply:
Now you see why the second one was created :)

At first I was trying to fix the old one, but the old one was never unit tested and it had some kind of "weird" patterns that I didn't dare to fix in order not to break backwards compatibility, therefore the second one was born

About the new hooks, as long as resume and suspend also gets called properly for sync actions (basically resume right after start and suspend right before finish) I don't see why not :)

Reply:
:) Iterating on the new api is definitely feeling more appealing now, I'll take a stab at it sometime this week.